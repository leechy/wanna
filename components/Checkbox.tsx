import { useThemeColor } from '@/hooks/useThemeColor';
import { StyleProp, StyleSheet, Text, TextStyle, TouchableOpacity, View, ViewStyle } from 'react-native';

import SquareIcon from '@/assets/symbols/square.svg';
import CheckIcon from '@/assets/symbols/square-check.svg';
import { useEffect, useState } from 'react';

interface CheckboxProps {
  checked: boolean;
  onChange?: (checked: boolean) => void;
  style?: StyleProp<ViewStyle>;
  labelStyle?: StyleProp<TextStyle>;
  color?: string;
  testID?: string;
  children?: React.ReactNode;
}

export default function Checkbox({ checked, onChange, style, labelStyle, color, testID, children }: CheckboxProps) {
  const primaryColor = useThemeColor({}, 'primary');
  const textColor = useThemeColor({}, 'text');

  const [localState, setLocalState] = useState<boolean>(checked);
  useEffect(() => {
    setLocalState(checked);
  }, [checked]);

  function updateState(newState: boolean) {
    setLocalState(newState);
    onChange?.(newState);
  }

  return (
    <TouchableOpacity
      onPressOut={() => updateState(!localState)}
      style={[styles.container, style]}
      testID={testID}
      activeOpacity={0.4}
    >
      <View style={styles.checkbox}>
        {localState ? (
          <CheckIcon width={28} height={28} color={color || primaryColor} />
        ) : (
          <SquareIcon width={28} height={28} color={color || primaryColor} />
        )}
      </View>
      <Text style={[styles.label, { color: textColor }, labelStyle]}>{children}</Text>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    marginVertical: 4,
    alignSelf: 'stretch',
  },
  checkbox: {
    marginRight: 8,
    alignSelf: 'flex-start',
  },
  label: {
    flex: 1,
    marginVertical: 3,
    fontFamily: 'Montserrat',
    fontSize: 16,
    lineHeight: 22,
  },
});
