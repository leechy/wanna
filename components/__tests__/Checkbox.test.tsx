import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Text } from 'react-native';
import Checkbox from '../Checkbox';

jest.mock('../../assets/symbols/square.svg', () => 'SquareIcon');
jest.mock('../../assets/symbols/square-check.svg', () => 'SquareCheckIcon');

describe('Checkbox', () => {
  it('renders correctly', () => {
    const { getByTestId } = render(
      <Checkbox checked={false} onChange={() => {}} testID="checkbox" />
    );
    expect(getByTestId('checkbox')).toBeTruthy();
  });

  it('calls onChange when clicked', () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <Checkbox checked={false} onChange={onChangeMock} testID="checkbox" />
    );
    const checkbox = getByTestId('checkbox');

    // Initially unchecked

    // Click to change state
    fireEvent.press(checkbox);
    expect(onChangeMock).toHaveBeenCalledWith(true);

    // Click again to change state
    fireEvent.press(checkbox);
    expect(onChangeMock).toHaveBeenCalledWith(false);
  });

  it('renders label with additional elements', () => {
    const { getByText, queryByText } = render(
      <Checkbox checked={false} onChange={() => {}} testID="checkbox">
        <Text>
          Label with{' '}
          <Text onPress={() => {}} testID="link">
            link
          </Text>
        </Text>
      </Checkbox>
    );

    // expect(queryByText('Label with')).toBeTruthy();
    expect(getByText('link')).toBeTruthy();
  });

  it('matches snapshot', () => {
    const { toJSON } = render(
      <Checkbox checked={false} onChange={() => {}} testID="checkbox">
        <Text>
          Label with{' '}
          <Text onPress={() => {}} testID="link">
            link
          </Text>
        </Text>
      </Checkbox>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
