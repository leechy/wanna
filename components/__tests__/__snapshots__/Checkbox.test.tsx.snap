// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "alignSelf": "stretch",
      "flexDirection": "row",
      "marginVertical": 4,
      "opacity": 1,
    }
  }
  testID="checkbox"
>
  <View
    style={
      {
        "alignSelf": "flex-start",
        "marginRight": 8,
      }
    }
  >
    <SquareIcon
      color="#3a1fc1"
      height={28}
      width={28}
    />
  </View>
  <Text
    style={
      [
        {
          "flex": 1,
          "fontFamily": "Montserrat",
          "fontSize": 16,
          "lineHeight": 22,
          "marginVertical": 3,
        },
        {
          "color": "#333333",
        },
        undefined,
      ]
    }
  >
    <Text>
      Label with
       
      <Text
        onPress={[Function]}
        testID="link"
      >
        link
      </Text>
    </Text>
  </Text>
</View>
`;
